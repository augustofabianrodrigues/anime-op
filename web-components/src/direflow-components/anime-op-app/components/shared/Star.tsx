import React, { FC, Fragment } from 'react';

interface StarProps {
  fillRatio: number;
}

const HEX_COLOR_FILLED = '#F6E05E';
const HEX_COLOR_EMPTY = '#718096';

function generateId (): string {
  return Math.random().toString(36).substring(2) + (new Date()).getTime().toString(36);
}

function renderGradientOffsets(fillRatio: number) {
  if (fillRatio <= 0) {
    return <stop offset="1" stopColor={HEX_COLOR_EMPTY} />;
  }

  const beginOffsetChange = fillRatio + 0.0001;
  if (beginOffsetChange >= 1) {
    return <stop offset="1" stopColor={HEX_COLOR_FILLED} />;
  }

  return (
    <Fragment>
      <stop offset={fillRatio} stopColor={HEX_COLOR_FILLED} />
      <stop offset={beginOffsetChange} stopColor={HEX_COLOR_EMPTY} />
    </Fragment>
  );
}

const Star: FC<StarProps> = (props) => {
  const paintId = 's_' + generateId();

  return (
    <svg
      width="22"
      height="21"
      viewBox="0 0 22 21"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10.3803 2.15903L12.9553 7.37569L18.7136 8.21736L14.5469 12.2757L15.5303 18.009L10.3803 15.3007L5.23027 18.009L6.21361 12.2757L2.04694 8.21736L7.80527 7.37569L10.3803 2.15903Z"
        fill={`url(#${paintId})`}
      />
      <defs>
        <linearGradient
          id={paintId}
          x1="2.04694"
          y1="10.4924"
          x2="18.7136"
          y2="10.4924"
          gradientUnits="userSpaceOnUse"
        >
          {renderGradientOffsets(props.fillRatio)}
        </linearGradient>
      </defs>
    </svg>
  );
};

export default Star;
