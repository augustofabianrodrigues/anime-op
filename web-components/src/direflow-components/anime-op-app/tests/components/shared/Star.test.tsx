import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Star from '../../../components/shared/Star';

jest.mock('../../../utils/generateId');

it('matches snapshot fill ratio of 0.68', () => {
  const renderTree = renderer.create(<Star fillRatio={0.68} />).toJSON();
  expect(renderTree).toMatchSnapshot();
});

it('matches snapshot fill ratio of 0', () => {
  const renderTree = renderer.create(<Star fillRatio={0} />).toJSON();
  expect(renderTree).toMatchSnapshot();
});

it('matches snapshot fill ratio of 0.11', () => {
  const renderTree = renderer.create(<Star fillRatio={0.11} />).toJSON();
  expect(renderTree).toMatchSnapshot();
});

it('matches snapshot fill ratio of 1', () => {
  const renderTree = renderer.create(<Star fillRatio={1} />).toJSON();
  expect(renderTree).toMatchSnapshot();
});
