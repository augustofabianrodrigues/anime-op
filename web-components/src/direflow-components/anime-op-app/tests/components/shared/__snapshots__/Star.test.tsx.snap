// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot fill ratio of 0 1`] = `
<svg
  fill="none"
  height="21"
  viewBox="0 0 22 21"
  width="22"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M10.3803 2.15903L12.9553 7.37569L18.7136 8.21736L14.5469 12.2757L15.5303 18.009L10.3803 15.3007L5.23027 18.009L6.21361 12.2757L2.04694 8.21736L7.80527 7.37569L10.3803 2.15903Z"
    fill="url(#s_undefined)"
  />
  <defs>
    <linearGradient
      gradientUnits="userSpaceOnUse"
      id="s_undefined"
      x1="2.04694"
      x2="18.7136"
      y1="10.4924"
      y2="10.4924"
    >
      <stop
        offset="1"
        stopColor="#718096"
      />
    </linearGradient>
  </defs>
</svg>
`;

exports[`matches snapshot fill ratio of 0.11 1`] = `
<svg
  fill="none"
  height="21"
  viewBox="0 0 22 21"
  width="22"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M10.3803 2.15903L12.9553 7.37569L18.7136 8.21736L14.5469 12.2757L15.5303 18.009L10.3803 15.3007L5.23027 18.009L6.21361 12.2757L2.04694 8.21736L7.80527 7.37569L10.3803 2.15903Z"
    fill="url(#s_undefined)"
  />
  <defs>
    <linearGradient
      gradientUnits="userSpaceOnUse"
      id="s_undefined"
      x1="2.04694"
      x2="18.7136"
      y1="10.4924"
      y2="10.4924"
    >
      <stop
        offset={0.11}
        stopColor="#F6E05E"
      />
      <stop
        offset={0.1101}
        stopColor="#718096"
      />
    </linearGradient>
  </defs>
</svg>
`;

exports[`matches snapshot fill ratio of 0.68 1`] = `
<svg
  fill="none"
  height="21"
  viewBox="0 0 22 21"
  width="22"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M10.3803 2.15903L12.9553 7.37569L18.7136 8.21736L14.5469 12.2757L15.5303 18.009L10.3803 15.3007L5.23027 18.009L6.21361 12.2757L2.04694 8.21736L7.80527 7.37569L10.3803 2.15903Z"
    fill="url(#s_undefined)"
  />
  <defs>
    <linearGradient
      gradientUnits="userSpaceOnUse"
      id="s_undefined"
      x1="2.04694"
      x2="18.7136"
      y1="10.4924"
      y2="10.4924"
    >
      <stop
        offset={0.68}
        stopColor="#F6E05E"
      />
      <stop
        offset={0.6801}
        stopColor="#718096"
      />
    </linearGradient>
  </defs>
</svg>
`;

exports[`matches snapshot fill ratio of 1 1`] = `
<svg
  fill="none"
  height="21"
  viewBox="0 0 22 21"
  width="22"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M10.3803 2.15903L12.9553 7.37569L18.7136 8.21736L14.5469 12.2757L15.5303 18.009L10.3803 15.3007L5.23027 18.009L6.21361 12.2757L2.04694 8.21736L7.80527 7.37569L10.3803 2.15903Z"
    fill="url(#s_undefined)"
  />
  <defs>
    <linearGradient
      gradientUnits="userSpaceOnUse"
      id="s_undefined"
      x1="2.04694"
      x2="18.7136"
      y1="10.4924"
      y2="10.4924"
    >
      <stop
        offset="1"
        stopColor="#F6E05E"
      />
    </linearGradient>
  </defs>
</svg>
`;
